### PR TITLE
[ctx_prof] Extend `WorkloadImportsManager` to use the contextual profile

### PR DESCRIPTION
**Pull request by mtrofin**
Keeping the json-based input as it's useful for diagnostics or for driving the import by other means than contextual composition.

The support for the contextual profile is just another modality for constructing the import list (`WorkloadImportsManager::Workloads`). Everything else - i.e. the actual importing logic - is already independent from how that list was obtained.

Issue #89287

### CODE REVIEW

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 1]
```diff
@@ -174,6 +174,10 @@ static cl::opt<std::string> WorkloadDefinitions(
              "}"),
     cl::Hidden);
 
+static cl::opt<std::string>
+    ContextualProfile("thinlto-pgo-ctx-prof",
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:178]: Can you add a test to exercise this handling?

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 2]
```diff
@@ -672,14 +670,78 @@ class WorkloadImportsManager : public ModuleImportsManager {
...
+      }
+      StringRef RootDefiningModule =
+          RootVI.getSummaryList().front()->modulePath();
+      LLVM_DEBUG(dbgs() << "[Workload] Root defining module for " << RootGuid
+                        << " is : " << RootDefiningModule << "\n");
+      auto &Set = Workloads[RootDefiningModule];
+      Root.getContainedGuids(ContainedGUIDs);
+      for (auto Guid : ContainedGUIDs)
+        if (auto VI = Index.getValueInfo(Guid))
+          Set.insert(VI);
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:717]: The json case has some debug printing of the eventual Set contents - needed here too?

File: llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp [hunk 3]
```diff
@@ -115,6 +115,15 @@ TEST_F(PGOCtxProfRWTest, RoundTrip) {
     EXPECT_EQ(Ctxes.size(), 2U);
     for (auto &[G, R] : roots())
       checkSame(*R, Ctxes.find(G)->second);
+
+    DenseSet<GlobalValue::GUID> Guids;
+    Ctxes.at(1U).getContainedGuids(Guids);
```
Comment (teresajohnson) [inline llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp:120]: It is unclear to me how this test change is related to the rest of the PR.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 2 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:717]: Moved in a common place.

File: llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp (hunk 3 above)
Comment (mtrofin) [reply llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp:120]: It's not, but this is the API that the rest of this patch hinges on. I could introduce the test separately.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:178]: We don't currently have a textual representation of the contextual profile, and the import manager is internal to this file, meaning we can't write a unittest. I'd do either, though, if the new functionality was more involved, but it's really just taking data from the ctx profile reader - which is tested - and filling a set with it.

I refactored a bit more to leave the specific code to a minimum.

File: llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp (hunk 3 above)
Comment (teresajohnson) [reply llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp:120]: I would just commit that separately (no need for another review, that part lgtm) since that isn't a new API here.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:178]: I meant a lit test. I think it would be good to add (reading the binary format like I am assuming the ctx profile reader tests do?), otherwise this is untested functionality.

File: llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp (hunk 3 above)
Comment (mtrofin) [reply llvm/unittests/ProfileData/PGOCtxProfReaderWriterTest.cpp]: done (a742693f6104055ec026852a70a68275fb82f7a0)

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:178]: Right - need to figure out where to place it, because it would need to be a test that runs a program to generate a profile. Meaning it needs compiler-rt.

Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:189]: done.

File: llvm/test/ThinLTO/X86/ctxprof.ll [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+; GUIDS-1-SAME: guid = 6019442868614718803
+; GUIDS-2: name: "m2_f1"
+; GUIDS-2-SAME: guid = 15593096274670919754
+;
+; RUN: rm -rf %t_baseline
+; RUN: rm -rf %t_exp
+; RUN: mkdir -p %t_baseline
+; RUN: mkdir -p %t_exp
+;
+; Normal run. m1 shouldn't get m2_f1 because it's not referenced from there.
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/ctxprof.ll:27]: and ditto for m2 and m1_f1?

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 5]
```diff
@@ -174,6 +174,17 @@ static cl::opt<std::string> WorkloadDefinitions(
              "}"),
     cl::Hidden);
 
+static cl::opt<bool> ImportAssumeUniqueLocal(
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:177]: Unrelated change should be removed

**Review by teresajohnson — APPROVED** [review]
lgtm with a couple small fixes

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 5 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:177]: this is from #100448?

File: llvm/test/ThinLTO/X86/ctxprof.ll (hunk 4 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/ctxprof.ll:27]: yup

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 5 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: Oh sorry, I was looking at diffs from my last review, and wasn't expecting things from other patches to show up (did it get rebased?)

Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: Yes, I needed to so I could get the llvm-ctxprof-util tool
